### PR TITLE
Djt carbon

### PR DESCRIPTION
**Pull request by Jacob8399**
This is the merge to include the carbon cycle in the master branch. Almost all of the changes happen in carbon_cycle.py, soil.py, and the output files for carbon. Carbon takes in residue from yield and outputs org_C% for RuFaS to use.

The model for carbon is fully functional and has room for advancements to be made. Certain variable changes will need to be addressed at a different time, whether they need to come from a database (silt clay), or they need to be reassed for carbon (fr_N). There are TODOs marking these variables.

### CODE REVIEW

File: RUFAS/output_handler/reports/field_report.py [hunk 1]
```diff
@@ -181,6 +183,134 @@ def __init__(self, data, field_name):
...
+                                    # 'metabolic_AG_L3': ['soil.soil_layers[2].metabolic_AG', 'kg/ha', []],
+                                    # 'metabolic_BG_L1': ['soil.soil_layers[0].metabolic_BG', 'kg/ha', []],
+                                    # 'metabolic_BG_L2': ['soil.soil_layers[1].metabolic_BG', 'kg/ha', []],
+                                    # 'metabolic_BG_L3': ['soil.soil_layers[2].metabolic_BG', 'kg/ha', []],
+                                    # 'structural_AG_L1': ['soil.soil_layers[0].structural_AG', 'kg/ha', []],
+                                    # 'structural_AG_L2': ['soil.soil_layers[1].structural_AG', 'kg/ha', []],
+                                    # 'structural_AG_L3': ['soil.soil_layers[2].structural_AG', 'kg/ha', []],
+                                    # 'structural_BG_L1': ['soil.soil_layers[0].structural_BG', 'kg/ha', []],
+                                    # 'structural_BG_L2': ['soil.soil_layers[1].structural_BG', 'kg/ha', []],
+                                    # 'structural_BG_L3': ['soil.soil_layers[2].structural_BG', 'kg/ha', []],
```
Comment (wmaxdonovan) [inline RUFAS/output_handler/reports/field_report.py]: This was part of the inspiration for the layer report structure in dm_manure. Do we want to consolidate this?

**Review by wmaxdonovan — CHANGES_REQUESTED** [review]
I think we should finish the pseudocode and consolidate variable names in accordance with RuFaS conventions / further break up some of these methods into smaller sections for readability prior to merging.

File: RUFAS/output_handler/reports/field_report.py (hunk 1 above)
Comment (Jacob8399) [reply RUFAS/output_handler/reports/field_report.py]: I'm actually not sure what you're talking about so if you think it should be consolidated then that should be fine. I'm definitely not too attached to the carbon output file haha.

**Comment by Jacob8399** [comment]
> I think we should finish the pseudocode and consolidate variable names in accordance with RuFaS conventions / further break up some of these methods into smaller sections for readability prior to merging.

Yeah that was the original goal but splitting everything up started to get messy and cause problems and I didn't want to change anything essential around the annual meeting. Personally, I agree it should be done but think we should do it in the patches branch just so we can get carbon and then dm_manure into master.

File: RUFAS/routines/soil/carbon_cycle/carbon_cycle.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+"""
+
+
+def update_all():
+    """
+    Description:
+        This function calls all the necessary functions to update information related
+        to the carbon cycle. The order in which each method is called is significant.
+
+        #TODO: Not yet implemented
```
Comment (HectorMM3) [inline RUFAS/routines/soil/carbon_cycle/carbon_cycle.py]: I don't understand why this has not been implemented yet?

File: RUFAS/routines/soil/soil.py [hunk 3]
```diff
@@ -184,9 +185,12 @@ def daily_soil_routine(soil, crop, weather, time):
 
     # calculate and update the contents of 3 organic and 2 inorganic nitrogen
     # pools
-    nitrogen_cycling.update_all(soil, weather, time)
+    nitrogen_cycle.update_all(soil, weather, time)
 
-    phosphorus_cycling.update_all(soil, weather, time)
+    phosphorus_cycle.update_all(soil, weather, time)
+
+    # TODO: Not currently implemented
```
Comment (HectorMM3) [inline RUFAS/routines/soil/soil.py]: @Jacob8399 what is the significance of this "TO DO?"

**Review by HectorMM3 — CHANGES_REQUESTED** [review]
Jacob, I left 2-3 years. about these commits. Can you please comment?

File: RUFAS/output_handler/reports/field_report.py (hunk 1 above)
Comment (HectorMM3) [reply RUFAS/output_handler/reports/field_report.py]: Let's discuss this on Th.

File: RUFAS/routines/field/crop/yields.py [hunk 4]
```diff
@@ -219,6 +219,12 @@ def calc_residue(soil, crop_type, field_management, time):
         field_management
         time
     """
+    # for carbon, needs to be calculated only at harvest
+    crop_type.bio_BG = crop_type.fr_root * crop_type.biomass_actual
+    soil.soil_layers[0].fr_tillage = 0.55
+    soil.lignin_residue_percent = 17
```
Comment (HectorMM3) [inline RUFAS/routines/field/crop/yields.py:216]: Eventually, this "17" will be removed.

File: RUFAS/routines/soil/carbon_cycle/carbon_cycle.py (hunk 2 above)
Comment (Jacob8399) [reply RUFAS/routines/soil/carbon_cycle/carbon_cycle.py]: I believe you're looking at outdated code.

File: RUFAS/routines/soil/soil.py (hunk 3 above)
Comment (Jacob8399) [reply RUFAS/routines/soil/soil.py]: Same as the above comment.

File: RUFAS/routines/field/soil/carbon_cycling/carbon_cycle.py [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+
+    Args:
+        soil
+        crop_type
+        weather
+        time
+    """
+
+    # S.6.A.1
+    d_lignin_residue_percent = 0.12 * weather.rainfall[time.year - 1][time.day - 1] * 0.01
```
Comment (HectorMM3) [inline RUFAS/routines/field/soil/carbon_cycling/carbon_cycle.py]: 0.01 should be 0.1, i.e., there is 0.1 mm per 1 cm. I rechecked the models and my vensim model had it wrong.

File: RUFAS/routines/field/soil/carbon_cycling/carbon_cycle.py [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+        struct_AG_to_BG = layer.structural_AG * layer.fr_tillage
+
+        # S.6.A.12
+        d_structural_AG = ((soil.residue_DM_harvest * (1 - metabolic_AG_frac)) - struct_AG_to_BG) - \
+                          (layer.struct_AG_to_C_active + layer.struct_AG_to_C_slow)
+        layer.structural_AG += d_structural_AG
+
+        # below ground metabolic residue and roots
+        # S.6.A.13
+        residue_DM_incorp = layer.fr_tillage * soil.residue_DM_harvest
```
Comment (HectorMM3) [inline RUFAS/routines/field/soil/carbon_cycling/carbon_cycle.py]: residue_DM_incorp name should be changed to residue_incorp

File: RUFAS/routines/field/soil/carbon_cycling/carbon_cycle.py [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+
+        # S.6.A.14
+        fr_lignin_residue_DM = 0
+        if residue_DM_incorp + crop_type.bio_BG != 0:
+            fr_lignin_residue_DM = residue_DM_incorp / (residue_DM_incorp + crop_type.bio_BG)
+
+        # S.6.A.15
+        soil.lignin_residue_BG_percent = soil.lignin_residue_BG_percent - 0.15 \
```
Comment (HectorMM3) [inline RUFAS/routines/field/soil/carbon_cycling/carbon_cycle.py]: Change 0.01 to 0.1

Comment (HectorMM3) [reply RUFAS/routines/field/soil/carbon_cycling/carbon_cycle.py]: This S.6.A.15 should not be allowed to go below zero.

File: RUFAS/routines/field/soil/carbon_cycling/carbon_cycle.py [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+        layer.structural_AG += d_structural_AG
+
+        # below ground metabolic residue and roots
+        # S.6.A.13
+        residue_DM_incorp = layer.fr_tillage * soil.residue_DM_harvest
+
+        # S.6.A.14
+        fr_lignin_residue_DM = 0
+        if residue_DM_incorp + crop_type.bio_BG != 0:
+            fr_lignin_residue_DM = residue_DM_incorp / (residue_DM_incorp + crop_type.bio_BG)
```
Comment (HectorMM3) [inline RUFAS/routines/field/soil/carbon_cycling/carbon_cycle.py]: Get rid of "DM"

File: RUFAS/routines/field/soil/carbon_cycling/carbon_cycle.py [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+
+        # S.6.A.15
+        soil.lignin_residue_BG_percent = soil.lignin_residue_BG_percent - 0.15 \
+                                         * weather.rainfall[time.year - 1][time.day - 1] * 0.01
+
+        # S.6.A.16
+        LN_ratio_BG = 0
+        if crop_type.fr_N != 0:
+            LN_ratio_BG = LN_ratio_AG * fr_lignin_residue_DM + ((soil.lignin_residue_percent / 100) / crop_type.fr_N) \
```
Comment (HectorMM3) [inline RUFAS/routines/field/soil/carbon_cycling/carbon_cycle.py]: get rid of DM

File: RUFAS/routines/field/soil/carbon_cycling/carbon_cycle.py [hunk 10]
```diff
@@ -0,0 +1,325 @@
...
+        layer.struct_BG_to_active_actual = struct_BG_to_active_actual
+
+        layer.struct_BG_to_slow_loss = layer.struct_BG_to_C_slow * fr_CO2_struct_to_slow
+        struct_BG_to_slow_actual = layer.struct_BG_to_C_slow * (1 - fr_CO2_struct_to_slow)
+        layer.struct_BG_to_slow_actual = struct_BG_to_slow_actual
+
+        # partitioning The Active and Slow Carbon Pools (in soil) Decomposition to Alternative Carbon Pools
+        # (e.g., Active Carbon Pool to Slow Carbon Pool) or Gas Loss
+
+        K5 = 0.14
```
Comment (HectorMM3) [inline RUFAS/routines/field/soil/carbon_cycling/carbon_cycle.py:212]: There should probably be a heading for K5 through K7 indicating that it is the active, slow, and passive carbon decomposition.

File: RUFAS/routines/field/soil/carbon_cycling/carbon_cycle.py [hunk 11]
```diff
@@ -0,0 +1,325 @@
...
+        soil
+    """
+
+    for layer in soil.soil_layers:
+        # soil mass calculations
+        # S.6.C.1
+        depth = layer.bottom_depth - soil.curr_layer_depth
+        soil.curr_layer_depth += depth
+
+        soil_volume = depth * 10 * soil.area
```
Comment (HectorMM3) [inline RUFAS/routines/field/soil/carbon_cycling/carbon_cycle.py:293]: Hi Jacob, where is this "10" coming from? meters*10*hectares...?

**Review by HectorMM3 — CHANGES_REQUESTED** [review]
Hi Jacob, I think in total things look good we need to go ahead and get rid of all "DM" designations since that needs to be removed from crop and the corresponding carbon code and pseudocode. Additionally, the mm to cm conversion we were using was 0.01. I know that you and I had decided to go with this value because it allowed us to match the vensim model. However, the true conversion is 0.1. A change in this conversion factor causes the below ground lignin adjustment to go negative. This needs to be capped at a zero min value. This will allow the Python model to work.  Please reach out if you have any questions. After we submit this I would like to make sure that Michael condenses the pseudocode, per Max's comments through DJT PATCHES are the merge. Also, this will set us up for the additional structures you will need to add in this semester.

File: RUFAS/routines/field/soil/carbon_cycling/carbon_cycle.py (hunk 10 above)
Comment (wmaxdonovan) [reply RUFAS/routines/field/soil/carbon_cycling/carbon_cycle.py:212]: Added note in pseudocode.

File: RUFAS/routines/field/soil/carbon_cycling/carbon_cycle.py (hunk 11 above)
Comment (wmaxdonovan) [reply RUFAS/routines/field/soil/carbon_cycling/carbon_cycle.py:293]: @Jacob8399 this is the only unresolved comment

**Review by wmaxdonovan — APPROVED** [review]

File: RUFAS/routines/field/soil/carbon_cycling/carbon_cycle.py (hunk 11 above)
Comment (Jacob8399) [reply RUFAS/routines/field/soil/carbon_cycling/carbon_cycle.py:293]: From carbon pseudocode:
soil volume = depth of soil layer*10,000*area
Where depth of a soil layer (meters) and 10,000 is the square meters per hectare (m2 ha-1).

Depth of soil layer needs to be in meters but rufas depth is in mm, so to get it to meters the equation then needs to be:
soil volume = (depth(mm) / 1000) * 10000 * area
Then I just simplified it to the 10

Comment (HectorMM3) [reply RUFAS/routines/field/soil/carbon_cycling/carbon_cycle.py:293]: Okay please make sure this is noted @Jacob8399 in the pseudocode. Thanks! :)

**Review by HectorMM3 — APPROVED** [review]

**Pull request merged**